index.ts

```diff
@@ -0,0 +1,45 @@
+interface IUnderlyingMap<K, V> {
+	has(key: K): boolean
+	get(key: K): V | undefined
+	delete(key: K): boolean
+	set(key: K, value: V): void
+}
+
+interface IManagedMap<K, V, GR, U extends IUnderlyingMap<K, V>> {
+	has(key: K): boolean
+	get(key: K): GR
+	delete(key: K): boolean
+	set(key: K, value: V): void
+	getUnderlyingDataStructure(): U
+}
+
+type ValueFactory<K, V> = (key: K) => V
+
+function mapFactory<K, V, U extends IUnderlyingMap<K, V> = Map<K, V>>(
+	factory?: undefined,
+	map?: U,
+): IManagedMap<K, V, undefined | V, U>
+function mapFactory<K, V, U extends IUnderlyingMap<K, V> = Map<K, V>>(
+	factory: ValueFactory<K, V>,
+	map?: U,
+): IManagedMap<K, V, V, U>
+function mapFactory<K, V, U extends IUnderlyingMap<K, V> = Map<K, V>>(
+	factory?: ValueFactory<K, V>,
```

Comment:
Do we need both of these argument type definitions if `factory` is no longer optional?